tests/datasets/test_streaming.py

```diff
@@ -0,0 +1,317 @@
+
+    # Build StreamingDataset
+    dataset = StreamingDataset(remote=remote, local=local, shuffle=shuffle, decoders=decoders, batch_size=batch_size)
+
+    # Test basic sample order
+    rcvd_samples = 0
+    shuffle_matches = 0
+    for ix, sample in enumerate(dataset):
+        rcvd_samples += 1
+        uid = sample["uid"]
+        data = sample["data"]
+        expected_uid = f"{ix:06}"
+        expected_data = 3 * ix
+        if shuffle:
+            shuffle_matches += (expected_uid == uid)
+        else:
+            assert uid == expected_uid == uid, f"sample ix={ix} has uid={uid}, expected {expected_uid}"
+            assert data == expected_data, f"sample ix={ix} has data={data}, expected {expected_data}"
+
+    # If shuffling, there should be few matches
+    # The probability of k matches in a random permutation is ~1/(e*(k!))
+    if shuffle:
+        assert shuffle_matches < 10
+
+    # Test length
+    assert rcvd_samples == num_samples, f"Only received {rcvd_samples} samples, expected {num_samples}"
+    assert len(dataset) == num_samples, f"Got dataset length={len(dataset)} samples, expected {num_samples}"
+
+
+@pytest.mark.timeout(10)
+@pytest.mark.parametrize("created_ago", [0.5, 3])
+@pytest.mark.parametrize("timeout", [1])
+def test_reader_after_crash(tmpdir: py.path.local, created_ago: float, timeout: float):
+    num_samples = 117
+    shard_size_limit = 1 << 8
+    samples, decoders = get_fake_samples_decoders(num_samples)
+    remote = str(tmpdir.mkdir("remote"))
+    local = str(tmpdir.mkdir("local"))
+    write_synthetic_streaming_dataset(dirname=remote, samples=samples, shard_size_limit=shard_size_limit)
+
+    shutil.copy(os.path.join(remote, "index.mds"), os.path.join(local, "index.mds.tmp"))
+    shutil.copy(os.path.join(remote, "000003.mds"), os.path.join(local, "000003.mds.tmp"))
+    time.sleep(created_ago)
+    dataset = StreamingDataset(remote=remote, local=local, shuffle=False, decoders=decoders, timeout=timeout)
+
+    # Iterate over dataset and make sure there are no TimeoutErrors
+    for _ in dataset:
+        pass
+
+
+@pytest.mark.parametrize(
+    "share_remote_local",
+    [
+        True,
+        pytest.param(False, marks=pytest.mark.xfail(reason="__getitem__ currently expects shards to exist")),
+    ],
+)
+def test_reader_getitem(tmpdir: py.path.local, share_remote_local: bool):
+    num_samples = 117
+    shard_size_limit = 1 << 8
+    samples, decoders = get_fake_samples_decoders(num_samples)
+
+    remote = str(tmpdir.mkdir("remote"))
+    write_synthetic_streaming_dataset(dirname=remote, samples=samples, shard_size_limit=shard_size_limit)
+    if share_remote_local:
+        local = remote
+    else:
+        local = str(tmpdir.mkdir("local"))
+
+    # Build StreamingDataset
+    dataset = StreamingDataset(remote=remote, local=local, shuffle=False, decoders=decoders)
+
+    # Test retrieving random sample
+    try:
+        _ = dataset[17]
+    except Exception as e:
```

Comment:
No need to catch the exception -- it's fine to have it propagate (in fact, catching it like this will hide the stack trace).